File: app/src/main/java/com/autotune/arabic/engine/RecordingWriter.java
package com.autotune.arabic.engine;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * يكتب الصوت المُعالَج إلى ملف WAV على الجهاز.
 *
 * يُفتح الملف عند بدء التسجيل، وتُكتب العينات بشكل متدفق،
 * ثم عند الإيقاف يُحدَّث رأس WAV بالحجم الفعلي.
 */
public class RecordingWriter {

    private static final int SAMPLE_RATE   = AudioEngine.SAMPLE_RATE;
    private static final int CHANNELS      = 1;
    private static final int BITS          = 16;
    private static final int BYTE_RATE     = SAMPLE_RATE * CHANNELS * BITS / 8;
    private static final int BLOCK_ALIGN   = CHANNELS * BITS / 8;

    private File outputFile;
    private FileOutputStream fos;
    private long dataBytes = 0;
    private volatile boolean active = false;

    /**
     * يبدأ تسجيلاً جديداً في المجلد المحدد.
     * @return الملف الذي سيُكتب فيه التسجيل
     */
    public File start(File dir) throws IOException {
        if (active) stop();

        String ts = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(new Date());
        outputFile = new File(dir, "AutoTune_" + ts + ".wav");
        fos = new FileOutputStream(outputFile);
        dataBytes = 0;
        writeWavHeader(fos, 0); // نكتب رأساً مؤقتاً بحجم صفر
        active = true;
        return outputFile;
    }

    /** يُكتب بلوك عينات صوتية (short) في الملف */
    public void write(short[] samples, int count) {
        if (!active || fos == null) return;
        byte[] buf = new byte[count * 2];
        for (int i = 0; i < count; i++) {
            buf[i * 2]     = (byte) (samples[i] & 0xFF);
            buf[i * 2 + 1] = (byte) ((samples[i] >> 8) & 0xFF);
        }
        try {
            fos.write(buf, 0, buf.length);
            dataBytes += buf.length;
        } catch (IOException ignored) {}
    }

    /**
     * يُوقف التسجيل ويُحدّث رأس WAV بالحجم الصحيح.
     * @return الملف المكتمل، أو null إذا لم يكن هناك تسجيل
     */
    public File stop() {
        if (!active) return null;
        active = false;
        try {
            fos.flush();
            fos.close();
            // نُحدّث حقلَي الحجم في رأس WAV باستخدام RandomAccessFile
            try (RandomAccessFile raf = new RandomAccessFile(outputFile, "rw")) {
                // حجم RIFF الكلي = 36 + dataBytes
                raf.seek(4);
                writeInt32LE(raf, (int)(36 + dataBytes));
                // حجم data chunk
                raf.seek(40);
                writeInt32LE(raf, (int) dataBytes);
            }
        } catch (IOException ignored) {}
        return outputFile;
    }

    public boolean isActive() { return active; }

    // ─── مساعدات WAV ─────────────────────────────────────────────────

    private void writeWavHeader(FileOutputStream out, long dataSize) throws IOException {
        // RIFF chunk
        out.write(new byte[]{'R','I','F','F'});
        writeInt32LE(out, (int)(36 + dataSize));
        out.write(new byte[]{'W','A','V','E'});
        // fmt sub-chunk
        out.write(new byte[]{'f','m','t',' '});
        writeInt32LE(out, 16);         // حجم fmt chunk
        writeInt16LE(out, 1);          // PCM = 1
        writeInt16LE(out, CHANNELS);
        writeInt32LE(out, SAMPLE_RATE);
        writeInt32LE(out, BYTE_RATE);
        writeInt16LE(out, BLOCK_ALIGN);
        writeInt16LE(out, BITS);
        // data sub-chunk
        out.write(new byte[]{'d','a','t','a'});
        writeInt32LE(out, (int) dataSize);
    }

    private void writeInt32LE(FileOutputStream out, int v) throws IOException {
        out.write(v & 0xFF);
        out.write((v >> 8) & 0xFF);
        out.write((v >> 16) & 0xFF);
        out.write((v >> 24) & 0xFF);
    }

    private void writeInt16LE(FileOutputStream out, int v) throws IOException {
        out.write(v & 0xFF);
        out.write((v >> 8) & 0xFF);
    }

    private void writeInt32LE(RandomAccessFile raf, int v) throws IOException {
        raf.write(v & 0xFF);
        raf.write((v >> 8) & 0xFF);
        raf.write((v >> 16) & 0xFF);
        raf.write((v >> 24) & 0xFF);
    }
}

File: ios/AutoTuneArabic/Engine/RecordingWriter.swift
import Foundation
import AVFoundation

final class RecordingWriter {
    private var fileHandle: FileHandle?
    private var outputURL: URL?
    private var sampleCount: Int = 0
    static let sampleRate: Int32 = 44100

    var isActive: Bool { fileHandle != nil }

    func start(saveDir: URL) throws -> URL {
        let timestamp = Int(Date().timeIntervalSince1970)
        let filename = "AutoTune_\(timestamp).wav"
        let url = saveDir.appendingPathComponent(filename)

        // Create directory if needed
        try FileManager.default.createDirectory(at: saveDir, withIntermediateDirectories: true, attributes: nil)

        // Create the file
        FileManager.default.createFile(atPath: url.path, contents: nil, attributes: nil)
        let handle = try FileHandle(forWritingTo: url)

        // Write placeholder WAV header (44 bytes, sizes = 0 initially)
        var header = Data(capacity: 44)

        // RIFF chunk
        header.append(contentsOf: [0x52, 0x49, 0x46, 0x46]) // "RIFF"
        appendUInt32LE(&header, 0)                           // file size - 8 (placeholder)
        header.append(contentsOf: [0x57, 0x41, 0x56, 0x45]) // "WAVE"

        // fmt sub-chunk
        header.append(contentsOf: [0x66, 0x6D, 0x74, 0x20]) // "fmt "
        appendUInt32LE(&header, 16)                          // sub-chunk size = 16
        appendUInt16LE(&header, 1)                           // PCM = 1
        appendUInt16LE(&header, 1)                           // mono = 1
        appendUInt32LE(&header, UInt32(RecordingWriter.sampleRate)) // sample rate
        appendUInt32LE(&header, UInt32(RecordingWriter.sampleRate) * 2) // byte rate (sr * channels * bitsPerSample/8)
        appendUInt16LE(&header, 2)                           // block align (channels * bitsPerSample/8)
        appendUInt16LE(&header, 16)                          // bits per sample

        // data sub-chunk
        header.append(contentsOf: [0x64, 0x61, 0x74, 0x61]) // "data"
        appendUInt32LE(&header, 0)                           // data size (placeholder)

        handle.write(header)

        fileHandle = handle
        outputURL = url
        sampleCount = 0
        return url
    }

    func write(samples: [Float]) {
        guard let handle = fileHandle else { return }
        var data = Data(capacity: samples.count * 2)
        for sample in samples {
            let clamped = max(-1.0, min(1.0, sample))
            let int16Val = Int16(clamped * 32767.0)
            var le = int16Val.littleEndian
            withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
        }
        handle.write(data)
        sampleCount += samples.count
    }

    func stop() -> URL? {
        guard let handle = fileHandle, let url = outputURL else { return nil }

        let dataSize = UInt32(sampleCount * 2)          // 16-bit = 2 bytes per sample
        let fileSize = dataSize + 36                     // 44 byte header - 8 byte RIFF header

        // Patch file size at offset 4
        handle.seek(toFileOffset: 4)
        var fs = fileSize.littleEndian
        handle.write(Data(bytes: &fs, count: 4))

        // Patch data size at offset 40
        handle.seek(toFileOffset: 40)
        var ds = dataSize.littleEndian
        handle.write(Data(bytes: &ds, count: 4))

        handle.closeFile()
        fileHandle = nil
        outputURL = nil
        sampleCount = 0
        return url
    }

    // MARK: - Helpers

    private func appendUInt32LE(_ data: inout Data, _ value: UInt32) {
        var le = value.littleEndian
        withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
    }

    private func appendUInt16LE(_ data: inout Data, _ value: UInt16) {
        var le = value.littleEndian
        withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
    }
}
